### PR TITLE
167 Remove unnecessary gems in zip artifact

### PR DESCRIPTION
**Pull request by sgnn7**
This PR:
- Splits out dev/test gems into the test docker image
- Removes dev/test gems from zip artifact
- Prevents builds from inadvertently updating gem lockfiles
- Speeds up gem installation by making it run in multiple threads
- Reduces docker context size for local builds (speeds up builds)
- Speeds up nokogiri installation by depending on system libs
- Prevents test code from using non-image-built code
- Cleans up dockerfiles

Connected to #167

### CODE REVIEW

File: build.sh [hunk 1]
```diff
@@ -2,28 +2,39 @@
...
+    echo 'Bundling dependencies...'
+    bundle package --all --no-install
+
+    echo 'Removing the old zip...'
+    rm -rf /output/cyberark-conjur-service-broker_$(cat VERSION).zip
+
+    echo 'Zipping everything up...'
+    zip -r \
+      /output/cyberark-conjur-service-broker_$(cat VERSION).zip ./* \
+      -x vendor/bundle
```
Comment (izgeri) [inline build.sh:39]: what does the `-x` do?

File: build.sh [hunk 2]
```diff
@@ -2,28 +2,39 @@
...
-  --rm \
+docker run --rm \
+  -v "$(pwd):/output" \
   conjur-service-broker \
-  bash -c "
-    bundle pack --all
-    zip -r cyberark-conjur-service-broker_$(cat VERSION).zip ./*
+  bash -ec "
+    echo 'Bundling dependencies...'
+    bundle package --all --no-install
```
Comment (izgeri) [inline build.sh:31]: what does the additional `--no-install` do?

File: docker-compose.yml [hunk 3]
```diff
@@ -55,8 +55,6 @@ services:
     command: rails s -p 3000 -b 0.0.0.0
```
Comment (izgeri) [inline docker-compose.yml:55]: Maybe these are wanted for development, to be able to run the tests with your local code?

It may be ok to remove them here, but when we get into this repo next and are making changes, we may want a mode available that maps our local code into the images to run the tests

File: build.sh (hunk 1 above)
Comment (sgnn7) [reply build.sh:39]: Excludes the directory. `vendor/bundle` is where the gems are installed.

File: build.sh (hunk 2 above)
Comment (sgnn7) [reply build.sh:31]: This grabs all gems and caches them without installation. Installation is not needed since we already do it earlier and we also don't include the installed gems in the zip.

File: build.sh (hunk 1 above)
Comment (izgeri) [reply build.sh:39]: so the zip doesn't include the packaged gems?

it is a requirement that the ZIP includes the vendored gems so that customers installing the service broker from an airgapped environment can still install the service broker and all of its runtime dependencies appropriately

File: docker-compose.yml (hunk 3 above)
Comment (sgnn7) [reply docker-compose.yml:55]: I totally see your point but the change serves multiple subtle purposes:
- It's a limitation of snipping out the dev/test gems. Without removal of the mounted dir, the gemfile with both test and dev gems is mounted and bundler blows up since it can't find the gems it needs. 
- Without relying strictly on the built image, we don't really know what is being tested since your local env could be different from the image
- It's a tiny speedup since it just needs to launch the pre-built image

File: build.sh (hunk 1 above)
Comment (sgnn7) [reply build.sh:39]: > so the zip doesn't include the packaged gems?

It's a bit complicated (as I found out myself). You have:
- Bundle Cache (`vendor/cache`): This is where the sources of those gems are stored. If something is needed to be installed, this directory will substitute for rubygems.
- Bundle installation directory (`vendor/bundle`): This is where those gems are installed. This is needed in our dockerfile since it needs to run `rails`.

You can use the cache to reconstruct the installation (with a few caveats that you can find [here](https://bundler.io/v1.6/bundle_package.html))

tl;dr: The gems will be installable it's just that we are only including the source package rather than also including the "installed" versions of those packages.

Comment (izgeri) [reply build.sh:39]: ok, that sounds reasonable. what is tricky is finding a way to test this :) maybe I should try running the service broker in docker without the internet on and see if it works?

File: docker-compose.yml (hunk 3 above)
Comment (izgeri) [reply docker-compose.yml:55]: makes sense, but maybe we need a follow-up issue to provide a way for devs to run these tests with their local code in a realistic way. it's fine if this is big enough to be out of scope for this effort. it might be worth adding a note to the contributing that explains that when you run these tests, you're not using your local code unless you've built the image first.

File: build.sh (hunk 1 above)
Comment (izgeri) [reply build.sh:39]: ran it with my internet off, and the tests passed.

File: docker-compose.yml (hunk 3 above)
Comment (izgeri) [reply docker-compose.yml:55]: @sgnn7 can you please update the CONTRIBUTING with instructions to build before running the tests locally after making changes? with that, I think this PR is gtg
